clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -424,6 +477,11 @@ ProducerIdAndEpoch producerIdAndEpoch() {
     }
 
     synchronized public void maybeUpdateProducerIdAndEpoch(TopicPartition topicPartition) {
+        if (hasFatalError()) {
```

Comment:
@urbandan LMK if you think this is correct or not. The `hasStaleProducerIdAndEpoch` does mutate some internal state which I'm _assuming_ we want to prevent if we're in a fatal state.

Reply:
Looks like this updates the epoch and resets the sequence for inflight batches. I don't think it hurts to have this, but it may be good to confirm.